providers/implementations/ciphers/cipher_ascon128.c

```diff
@@ -0,0 +1,710 @@
+#  define OSSL_CIPHER_PARAM_AEAD_AAD "aad"
+# endif
+
+/* Note: get_iv_length and get_tag_length are not standard OpenSSL dispatch functions.
+ * IV and tag lengths are retrieved via get_ctx_params instead.
+ * These functions are kept for internal use only.
+ */
+
+/* ASCON-128 uses a fixed key length of 16 bytes (128 bits) */
+
+static void ascon128_cleanctx(void *vctx)
+{
+    struct ascon_ctx_st *ctx = vctx;
+
+    ctx->is_tag_set = false;
+    ctx->is_ongoing = false;
+    ctx->assoc_data_processed = false;
+    ctx->tag_len = FIXED_TAG_LENGTH;
+    ctx->iv_set = false;
+    memset(ctx->internal_ctx, 0, sizeof(*(ctx->internal_ctx)));
+    memset(ctx->tag, 0, sizeof(ctx->tag));
+    memset(ctx->iv, 0, sizeof(ctx->iv));
+}
+
+static void *ascon128_newctx(void *provctx)
+{
+    struct ascon_ctx_st *ctx;
+    intctx_t *intctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL)
+        return NULL;
+
+    ctx->provctx = provctx;
+    ctx->is_tag_set = false;
+    ctx->is_ongoing = false;
+    ctx->assoc_data_processed = false;
+    ctx->tag_len = FIXED_TAG_LENGTH;  /* default tag length */
+    ctx->iv_set = false;
+
+    intctx = OPENSSL_zalloc(sizeof(*intctx));
+    if (intctx == NULL)
+    {
+        OPENSSL_clear_free(ctx, sizeof(*ctx));
+        return NULL;
+    }
+    ctx->internal_ctx = intctx;
+
+    return ctx;
+}
+
+static void *ascon128_dupctx(void *vctx)
+{
+    struct ascon_ctx_st *src = vctx;
+    struct ascon_ctx_st *dst = NULL;
+
+    if (src == NULL || !ossl_prov_is_running())
+        return NULL;
+
+    /* Create new context using the same provider context */
+    if ((dst = ascon128_newctx(src->provctx)) == NULL)
+        return NULL;
+
+    /* Copy all context fields */
+    dst->direction = src->direction;
+    dst->is_ongoing = src->is_ongoing;
+    dst->is_tag_set = src->is_tag_set;
+    dst->assoc_data_processed = src->assoc_data_processed;
+    dst->tag_len = src->tag_len;
+    dst->iv_set = src->iv_set;
+
+    /* Copy tag if it's set */
+    if (src->is_tag_set) {
+        memcpy(dst->tag, src->tag, FIXED_TAG_LENGTH);
+    }
+
+    /* Copy IV if it's set */
+    if (src->iv_set) {
+        memcpy(dst->iv, src->iv, ASCON_AEAD_NONCE_LEN);
+    }
+
+    /* Deep copy the internal LibAscon context */
+    if (src->internal_ctx != NULL && dst->internal_ctx != NULL) {
+        memcpy(dst->internal_ctx, src->internal_ctx, sizeof(*dst->internal_ctx));
+    }
+
+    return dst;
+}
+
+static void ascon128_freectx(void *vctx)
+{
+    struct ascon_ctx_st *ctx = vctx;
+
+    if (ctx == NULL)
+        return;
+
+    ctx->provctx = NULL;
+    ascon128_cleanctx(ctx);
+    OPENSSL_clear_free(ctx->internal_ctx, sizeof(*ctx->internal_ctx));
+    OPENSSL_clear_free(ctx, sizeof(*ctx));
+}
+
+/* Internal initialization function (shared by encrypt and decrypt init) */
+
+static int ascon128_internal_init(void *vctx, direction_t direction,
```

Comment:
There are some big differences between what it handles here and what AES GCM allows.
i.e AES GCM allows key and iv to be set on multiple calls.

Also AES_GCM handles AAD as part of its update(), do we want to diverge from that here?

Long term we should have a AEAD API, but until then should we allow this to be done differently?

@openssl/committers this is a design decision..